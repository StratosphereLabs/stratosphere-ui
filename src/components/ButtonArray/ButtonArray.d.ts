import { ComponentProps, FC, HTMLProps, Key, ReactNode } from '../../../node_modules/react';
import { Breakpoint } from '../../common';
import { ButtonProps } from '../Button';
import { DropdownMenuProps } from '../DropdownMenu';
export type ButtonOptions = Omit<ButtonProps, 'onClick'> & {
    icon: FC<ComponentProps<'svg'>>;
    key: Key;
    menuText?: string;
    onClick: () => void;
    tooltipText?: string;
};
export interface ButtonArrayProps extends Omit<HTMLProps<HTMLDivElement>, 'children'> {
    buttonOptions: ButtonOptions[];
    collapseAt?: Breakpoint;
    dropdownMenuProps?: Omit<DropdownMenuProps, 'items'>;
    withTooltips?: boolean;
}
export interface ButtonWrapperProps {
    children: ReactNode;
    tooltipText?: string;
}
export declare const ButtonArray: ({ buttonOptions, collapseAt, dropdownMenuProps: { buttonProps: dropdownButtonProps, ...dropdownMenuProps }, withTooltips, ...props }: ButtonArrayProps) => JSX.Element;
