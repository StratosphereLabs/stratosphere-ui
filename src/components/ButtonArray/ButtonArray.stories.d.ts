import { Meta, StoryObj } from '@storybook/react';
import { ButtonArray } from './ButtonArray';
declare const meta: Meta<typeof ButtonArray>;
export default meta;
type Story = StoryObj<typeof ButtonArray>;
export declare const Primary: Story;
export declare const WithTooltips: Story;
export declare const Collapsable: Story;
