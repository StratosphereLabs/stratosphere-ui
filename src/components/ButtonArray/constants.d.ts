import { Breakpoint } from '../../common';
export declare const FLEX_BREAKPOINTS_MAP: Record<Breakpoint, string>;
export declare const HIDDEN_BREAKPOINTS_MAP: Record<Breakpoint, string>;
