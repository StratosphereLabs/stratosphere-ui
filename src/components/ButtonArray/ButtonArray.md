# ButtonArray

The `ButtonArray` component is a reusable React component that renders an array of Buttons and can collapse into a Dropdown menu. It accepts various props to modify its appearance and behavior.

## Props

The `ButtonArray` component accepts the following props:

- `buttonOptions (required)`: An array of objects representing the individual buttons and their properties. Each object should have the following properties:
  - `children`: (ReactNode) The content of the button.
  - `className`: (string) Additional CSS class for the button.
  - `icon`: (React Component) The icon component to be displayed.
  - `key`: (Key) A unique key for the button.
  - `menuText`: (string) Text to be displayed in the button's dropdown menu.
  - `onClick`: (function) Callback function to be executed when the button is clicked.
    Additional props from `ButtonProps` are also accepted (excluding `onClick`).
  - `tooltipText`: (string) An optional string that customizes the text for the tooltip. Defaults to `menuText`.
- `className`: (string) Additional CSS class for the component.
- `collapseAt`: (string) Screen size breakpoint at which buttons collapse into a dropdown menu ('sm', 'md', 'lg', 'xl').
- `dropdownMenuProps`: (object) Props to customize the dropdown menu:
- `withTooltips`: (boolean) An optional boolean that enables tooltips when hovering over buttons.

## Usage

```tsx
import { ButtonArray } from 'stratosphere-ui';

const MyComponent = () => {
  const buttonOptions = [
    {
      icon: MyIconComponent,
      key: 'button1',
      menuText: 'Button 1 Menu',
      onClick: () => {
        /* Handle button 1 click */
      },
    },
    {
      icon: AnotherIconComponent,
      key: 'button2',
      menuText: 'Button 2 Menu',
      onClick: () => {
        /* Handle button 2 click */
      },
    },
    // ... more button options
  ];

  return (
    <ButtonArray
      buttonOptions={buttonOptions}
      className="my-button-array"
      collapseAt="md"
      dropdownMenuProps={{
        menuSize: 'lg',
        // ... other dropdown menu props
      }}
    />
  );
};
```

## Notes

- When `collapseAt` is provided, the buttons will collapse into a dropdown menu when the screen width is below the specified breakpoint.
- The `buttonOptions` array defines the buttons and their behaviors. Each button can have an optional dropdown menu.
- The `icon` property in each `buttonOptions` object should be a valid React component representing an SVG icon.
