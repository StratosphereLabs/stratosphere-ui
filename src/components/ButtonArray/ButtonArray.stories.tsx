import { Story, Meta } from '@storybook/react';
import { ButtonArray, ButtonArrayProps } from './ButtonArray';
import { EyeIcon, InfoIcon, WarningIcon } from '../Icons';

export default {
  title: 'ButtonArray',
  component: ButtonArray,
  argTypes: {
    icon: {
      control: false,
    },
  },
} as Meta;

export const Default: Story<ButtonArrayProps> = args => (
  <ButtonArray {...args} />
);

Default.args = {
  buttonOptions: [
    {
      color: 'accent',
      key: 'view',
      icon: EyeIcon,
      menuText: 'View',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'info',
      key: 'info',
      icon: InfoIcon,
      menuText: 'Info',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'warning',
      key: 'warning',
      icon: WarningIcon,
      menuText: 'Warning',
      onClick: () => null,
      size: 'xs',
    },
  ],
};

export const WithTooltips: Story<ButtonArrayProps> = args => (
  <ButtonArray {...args} />
);

WithTooltips.args = {
  buttonOptions: [
    {
      color: 'accent',
      key: 'view',
      icon: EyeIcon,
      menuText: 'View',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'info',
      key: 'info',
      icon: InfoIcon,
      menuText: 'Info',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'warning',
      key: 'warning',
      icon: WarningIcon,
      menuText: 'Warning',
      onClick: () => null,
      size: 'xs',
    },
  ],
  withTooltips: true,
};

export const Collapsable: Story<ButtonArrayProps> = args => (
  <ButtonArray {...args} />
);

Collapsable.args = {
  buttonOptions: [
    {
      color: 'accent',
      key: 'view',
      icon: EyeIcon,
      menuText: 'View',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'info',
      key: 'info',
      icon: InfoIcon,
      menuText: 'Info',
      onClick: () => null,
      size: 'xs',
    },
    {
      color: 'warning',
      key: 'warning',
      icon: WarningIcon,
      menuText: 'Warning',
      onClick: () => null,
      size: 'xs',
    },
  ],
  className: 'pb-[120px]',
  collapseAt: 'xl',
  withTooltips: true,
};
