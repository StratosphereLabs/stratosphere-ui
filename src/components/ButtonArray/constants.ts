import { Breakpoint } from '../../common';

export const FLEX_BREAKPOINTS_MAP: Record<Breakpoint, string> = {
  sm: 'sm:flex',
  md: 'md:flex',
  lg: 'lg:flex',
  xl: 'xl:flex',
};

export const HIDDEN_BREAKPOINTS_MAP: Record<Breakpoint, string> = {
  sm: 'sm:hidden',
  md: 'md:hidden',
  lg: 'lg:hidden',
  xl: 'xl:hidden',
};
