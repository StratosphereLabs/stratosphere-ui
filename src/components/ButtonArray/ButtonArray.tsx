import classNames from 'classnames';
import {
  type Key,
  type ComponentProps,
  type FC,
  Fragment,
  HTMLProps,
  ReactNode,
} from 'react';
import { Breakpoint } from '../../common';
import { Button, ButtonProps } from '../Button';
import { DropdownMenu, DropdownMenuProps } from '../DropdownMenu';
import { EllipsisVerticalIcon } from '../Icons';
import { FLEX_BREAKPOINTS_MAP, HIDDEN_BREAKPOINTS_MAP } from './constants';

export type ButtonOptions = Omit<ButtonProps, 'onClick'> & {
  icon: FC<ComponentProps<'svg'>>;
  key: Key;
  menuText?: string;
  onClick: () => void;
  tooltipText?: string;
};

export interface ButtonArrayProps
  extends Omit<HTMLProps<HTMLDivElement>, 'children'> {
  buttonOptions: ButtonOptions[];
  collapseAt?: Breakpoint;
  dropdownMenuProps?: Omit<DropdownMenuProps, 'items'>;
  withTooltips?: boolean;
}

export interface ButtonWrapperProps {
  children: ReactNode;
  tooltipText?: string;
}

export const ButtonArray = ({
  buttonOptions,
  collapseAt,
  dropdownMenuProps: {
    buttonProps: dropdownButtonProps,
    ...dropdownMenuProps
  } = {},
  withTooltips,
  ...props
}: ButtonArrayProps): JSX.Element => {
  const ButtonWrapper = withTooltips
    ? ({ children, tooltipText }: ButtonWrapperProps) => (
        <div className="tooltip" data-tip={tooltipText}>
          {children}
        </div>
      )
    : Fragment;
  return (
    <div {...props}>
      <div
        className={classNames(
          'gap-1',
          collapseAt === undefined && 'flex',
          collapseAt !== undefined &&
            `hidden ${FLEX_BREAKPOINTS_MAP[collapseAt]}`,
        )}
      >
        {buttonOptions.map(
          ({
            children,
            className: buttonClassName,
            icon: Icon,
            key,
            menuText,
            onClick,
            tooltipText,
            ...buttonProps
          }) => (
            <ButtonWrapper tooltipText={tooltipText ?? menuText} key={key}>
              <Button
                className={classNames('px-1', buttonClassName)}
                onClick={onClick}
                {...buttonProps}
              >
                <Icon className="h-4 w-4" />
                {children ?? <span className="sr-only">{menuText}</span>}
              </Button>
            </ButtonWrapper>
          ),
        )}
      </div>
      <div
        className={classNames(
          collapseAt === undefined && 'hidden',
          collapseAt !== undefined &&
            `flex ${HIDDEN_BREAKPOINTS_MAP[collapseAt]}`,
        )}
      >
        <DropdownMenu
          buttonProps={{
            color: 'ghost',
            shape: 'circle',
            size: 'sm',
            children: (
              <>
                <EllipsisVerticalIcon className="h-5 w-5" />
                <span className="sr-only">Actions Menu</span>
              </>
            ),
            ...dropdownButtonProps,
          }}
          items={buttonOptions.map(
            ({ icon: Icon, key, menuText, onClick }) => ({
              id: typeof key === 'number' ? key.toString() : key,
              onClick,
              children: (
                <>
                  <Icon className="h-4 w-4" />
                  {menuText}
                </>
              ),
            }),
          )}
          {...dropdownMenuProps}
        />
      </div>
    </div>
  );
};
