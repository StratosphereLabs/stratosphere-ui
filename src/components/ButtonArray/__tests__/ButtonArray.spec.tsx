import { fireEvent, render, screen } from '@testing-library/react';
import { FC } from 'react';

import { ButtonArray, ButtonOptions } from '../ButtonArray';

const MockIcon: FC = () => <svg data-testid="mock-icon" />;

class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

describe('ButtonArray Component', () => {
  window.ResizeObserver = ResizeObserver;

  const buttonOptions: ButtonOptions[] = [
    {
      key: 'button1',
      icon: MockIcon,
      onClick: vi.fn(),
      menuText: 'Button 1',
    },
    {
      key: 'button2',
      icon: MockIcon,
      onClick: vi.fn(),
      menuText: 'Button 2',
    },
  ];

  test('renders buttons when collapseAt is undefined', () => {
    render(<ButtonArray buttonOptions={buttonOptions} />);

    expect(screen.getByText('Button 1')).toBeInTheDocument();
    expect(screen.getByText('Button 2')).toBeInTheDocument();
  });

  test('calls onClick when button is clicked', () => {
    render(<ButtonArray buttonOptions={buttonOptions} />);

    const button1 = screen.getByText('Button 1');
    fireEvent.click(button1);

    expect(buttonOptions[0].onClick).toHaveBeenCalled();
  });

  test('renders tooltips when withTooltips is true', () => {
    render(<ButtonArray buttonOptions={buttonOptions} withTooltips />);

    expect(screen.getByText('Button 1')).toBeInTheDocument();
    expect(screen.getByText('Button 2')).toBeInTheDocument();
  });

  test('renders dropdown when collapseAt is set', () => {
    render(<ButtonArray buttonOptions={buttonOptions} collapseAt="sm" />);

    expect(screen.getByText('Actions Menu')).toBeInTheDocument();
  });

  test('dropdown button click triggers menu options', () => {
    render(<ButtonArray buttonOptions={buttonOptions} collapseAt="sm" />);

    fireEvent.click(screen.getByText('Actions Menu'));
    expect(screen.getAllByText('Button 1')[1]).toBeInTheDocument();
    expect(screen.getAllByText('Button 2')[1]).toBeInTheDocument();
  });
});
