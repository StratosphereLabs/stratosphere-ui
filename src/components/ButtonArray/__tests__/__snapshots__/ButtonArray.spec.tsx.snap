// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ButtonArray Component > renders tooltips when withTooltips is true 1`] = `
<div>
  <div>
    <div
      class="gap-1 flex"
    >
      <div
        class="tooltip"
        data-tip="Tooltip 1"
      >
        <button
          class="btn px-1"
          type="button"
        >
          <svg />
          <span
            class="sr-only"
          >
            Menu 1
          </span>
        </button>
      </div>
      <div
        class="tooltip"
        data-tip="Menu 2"
      >
        <button
          class="btn px-1"
          type="button"
        >
          <svg />
          <span
            class="sr-only"
          >
            Menu 2
          </span>
        </button>
      </div>
    </div>
    <div
      class="hidden"
    >
      <div
        class="relative"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          class="btn btn-ghost btn-circle btn-sm"
          data-headlessui-state=""
          id="headlessui-menu-button-:rb:"
          type="button"
        >
          <svg
            class="h-5 w-5"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 6.75a.75.75 0 110-1.5.75.75 0 010 1.5zM12 12.75a.75.75 0 110-1.5.75.75 0 010 1.5zM12 18.75a.75.75 0 110-1.5.75.75 0 010 1.5z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span
            class="sr-only"
          >
            Actions Menu
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ButtonArray Component > renders without crashing 1`] = `
<div>
  <div>
    <div
      class="gap-1 flex"
    >
      <button
        class="btn px-1"
        type="button"
      >
        <svg />
        <span
          class="sr-only"
        >
          Menu 1
        </span>
      </button>
      <button
        class="btn px-1"
        type="button"
      >
        <svg />
        <span
          class="sr-only"
        >
          Menu 2
        </span>
      </button>
    </div>
    <div
      class="hidden"
    >
      <div
        class="relative"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          class="btn btn-ghost btn-circle btn-sm"
          data-headlessui-state=""
          id="headlessui-menu-button-:r2:"
          type="button"
        >
          <svg
            class="h-5 w-5"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 6.75a.75.75 0 110-1.5.75.75 0 010 1.5zM12 12.75a.75.75 0 110-1.5.75.75 0 010 1.5zM12 18.75a.75.75 0 110-1.5.75.75 0 010 1.5z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span
            class="sr-only"
          >
            Actions Menu
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
